docs/compilers/CSharp/Nullability Public API Design Notes.md

```diff
@@ -0,0 +1,241 @@
+For these cases, flow state is necessary to compute the declared nullability.
+For example:
+
+```C#
+string? s1 = string.Empty;
+var s2 = s1;
+s2 = null; // This will warn, because the type of s2 is string, not string?
+```
+
+The declared nullability of `s1` is `MayBeNull`, even though it is being initialized with a non-null value.
+However, because of the flow state of `s1` is `NotNull`, we will infer the declared nullability of `s2` to be `NotNull` as well.
+
+Declared nullability can be retreived from the declaration symbol (e.g. `IFieldSymbol`, `IMethodSymbol`, etc.).
+
+### Flow State
+
+When the nullable feature is enabled, Roslyn will track the flow state of expressions throughout a method, regardless of what the variable was declared as.
+`SemanticModel` can be used to request flow state information for individual expressions or sub-expressions through the `GetTypeInfo` APIs.
+The `TypeInfo` struct will be augmented with `Nullability` and `ConvertedNullability` information, similar to how it currently has `Type` and `ConvertedType`.
+Nullability of symbols at a location can be retrieved through the `GetSymbolInfo` API: if the expression successfully resolved to a single symbol, the `Symbol` property will have nullability calculated.
+`CandidateSymbols` will not have calculated nullabilities, as we cannot run nullable analysis on code that has failed overload resolution.
+
+### `GetSemanticModel(bool skipNullabilityInformation)`
+
+This API proposal changes the way `SemanticModel` calls behave by default.
+Currently, if semantic information is requested for a `SyntaxNode`, we make an effort to bind as little as possible to provide semantic information about that statement, usually a single statement.
+However, if that semantic information will include `Nullability`, we must bind all statements that come before that statement and run nullability analysis up to and including that statement.
+We believe that this change will be ok for most scenarios, as things such as colorization and analyzers are already running in the background near-constantly, binding pretty much everything on screen and in the file anyway.
+There are a few specific scenarios that will be perf-sensitive, and an overload for `GetSemanticModel()` that skips nullability information will be provided for these.
+It will share an initial-binding `BoundNode` cache with the standard semantic model, so it will get the benefits if something else with the same semantic model requests information, as it would have today.
+
+## New APIs
+
+We add the new APIs for retreiving and interacting with nullability:
+
+```C#
+public enum Nullability
+{
+    NotApplicable,
+    NotComputed,
+    Unknown,
+    NotNull,
+    MayBeNull
+}
+
+#region Declared Nullability
+
+// These can potentially be influenced by calculated flow state, but do not change after initial declaration
+
+public interface IDiscardSymbol : ISymbol
+{
+    Nullability Nullability { get; }
```

Comment:
this seems weird, given that you can't really interact with a discarded symbol right?  So when does this Nullability ever come into play?

Reply:
You can't interact with it, but the IDE will need to display something for QuickInfo when the user hovers over it so we need to provide the info.